src/ucp/tag/tag_recv.c

```diff
@@ -63,7 +64,9 @@ ucp_tag_recv_common(ucp_worker_h worker, void *buffer, size_t count,
         req->recv.tag.info.sender_tag = ucp_rdesc_get_tag(rdesc);
         req->recv.tag.info.length     = recv_len;
 
-        status = ucp_dt_unpack_only(buffer, count, datatype,
+        ucp_memory_type_detect_mds(worker->context, (void *)buffer, recv_len, &mem_type);
```

Comment:
need to make this function inline to avoid the xtra overhead
also can you please check how this PR affects performance (osu_lat,osu_bw) of small messages?

Reply:
no need to cast to void*